Create the design for success principle (#63)
HO-CTO/engineering-guidance-and-standards: docs/principles/design-for-success.md

---
layout: side-navigation
order: 1
title: Design for success
date: git Last Modified
tags:
- Software design
---

Last updated: {{ page.date | postDate }}
Tags: {{ tags | join(', ') }}

> Systems and applications should be designed so that they can be efficiently developed and deliver their intended value. You should have success criteria to measure your designs against.

Reviewer: Mention users here

---

## Rationale

> A clear understanding and focus on meeting user and business needs is vital for delivering value for money when designing software solutions. Meeting non-functional requirements at the right time (often earlier) reduces the effort and cost of building secure (also check [secure by design](#)) and performant applications.

Author: We don't yet have a 'secure by design' page, so probably best not to add a link to it until present.

Author: I think remove everything in the brackets for now, once more guidance/principles/etc gets added then more links can be woven through existing docs

Ensuring that there are pragmatic success criteria for what is to be built, how and when, makes assurance of design easier, allows for better management of technical debt and carries through to more maintainable solutions

> 

Author: Separation inline suggestion made to incorporate mention of testing

---

## Applications and Implications

> - Engage with and understand the needs of your stakeholders. Get their feedback regularly to be assured that you are delivering value

Reviewer: Talk more about the users and not just stakeholders

Reviewer: Give examples of how you might achieve it

- Actively manage technical debt and design your implementation so the solution is the right size at the right time

> - Build or leverage reusable components and patterns where appropriate

Reviewer: Use existing languages and frameworks (within the HO)

> - Design for the minimal team - consider existing Home Office engineering capabilities when making technology choices so that iteration and support are achievable without unnecessary overhead

Reviewer: More explicit around adding new tech stacks/design choices etc. when we don't have the engineering capacity to support it in the future.

> - Use Wardley maps to inform designs - this will allow you to more confidently identify where you should innovate and where you can consume existing commodity components

Reviewer: Pick the times when you need to be innovative or 'get something off the shelf'. Where are your components coming from? Wardley maps help you do that more confidently

- Keep it simple, but think about extensibility - will your design allow for delivery of potential future needs? Can you use a simpler framework that will allow you to deliver more quickly?

> 

Reviewer: Remember that engineers are users too

> _These are guildelines, some may not be applicable, or used in certain circumstances_

Author: I think I'd prefer an introductory paragraph under the 'Applications and implications' title, perhaps something like:

> There are a number of things that you can do to improve the chances of success with your design:

Having the existing caveat at the bottom of the document could imply that it covers everything within it.

---
layout: sub-navigation
order: 1
title: Design for success
date: git Last Modified
tags:
  - Software design
---

{% import "_includes/macros/tags.njk" as pageTag %}

Last updated: {{ page.date | postDate }}
{{ pageTag.tags(tags)  }}

> Systems and applications should be designed so that they can be efficiently developed and deliver their intended value for users. You should have success criteria to measure your designs against.

Reviewer 2: This paragraph isn't enough for me to really understand the principle. I wonder if it would benefit from some bullet points that list the considerations for your project/service to be successful. _Something_ like:
- Meets user needs
- Meets business needs
- Meets security requirements
- Is reliable
- Adheres to governance (there will be a better way of phrasing this)
- Can be delivered to time and budget
- Is easy to maintain

Reviewer 2: Governance might include the service standard, PSBAR, GDPR & data protection, as well as working with TDA.

Reviewer: This is meant to be a one liner I think, then these points are meant to be covered in rationale and Applications, but maybe this can be incorporated elsewhere if you think we haven't covered them all?

Author: ```suggestion
Systems and applications should be designed so that they can be efficiently maintained, developed and deliver their intended value. Use success criteria to measure your designs against user needs, business and governance direction, and non-functional requirements such as reliability and security.
```

---

## Rationale

A clear understanding and focus on meeting user and business needs is vital for delivering value for money when designing software solutions. Meeting non-functional requirements at the right time (often earlier) reduces the effort and cost of building secure and performant applications.

> Ensuring that there are pragmatic success criteria for what is to be built, how and when, makes assurance of design easier, allows for better management of technical debt and carries through to more maintainable solutions

Author: ```suggestion
Ensuring that there are pragmatic and timely success criteria for what is to be built makes testing and assurance of design easier, allows for better management of technical debt and carries through to more maintainable solutions.
```

---

## Applications and Implications

There are a number of things that you can do to improve the chances of success with your design:

- Engage with and understand the needs of your users and stakeholders. Get their feedback regularly to be assured that you are delivering value (for example using user research, using prototyping and fast feedback)
- Actively manage technical debt and design your implementation so the solution is the right size at the right time
- Build or leverage reusable components and patterns where appropriate

> - Design for the minimal team - consider existing Home Office engineering capabilities when making technology choices (use languages and frameworks that the Home Office already has experience in where appropriate) so that iteration and support are achievable without unnecessary overhead. Remember that developers are users too.

Reviewer 2: When designing for a minimal team, I'd think more locally than org-wide. Org-wide alignment has benefits but local alignment has even greater ones. e.g. The Home Office has expertise in both Java and Node.js, but rather than using both in your service, you should try to use just one if at all possible. That way, you developers don't need to have expertise in both, and you avoid having 'teams within teams', where a ticket might be passed to a backend/API dev and then to a frontend/UI dev.

Org-wide alignment makes our recruitment easier and allows us to more easily move people between services.

Local alignment means that you can support the service with fewer people, which makes the service more sustainable in the long run. I think this is going to be very important as we move towards product-centricity and so want to have enduring teams rather than throwing services over the fence to a support team.

Author: I think there is definitely value in us creating some more detailed guidance around this point later, but for now I think Aaron's revised text covers the high-level and we should try to keep the principle as succinct as possible

- Pick the times when you need to be innovative or 'get something off the shelf'. Wardley maps help you do that more confidently

> - Keep it simple, but think about extensibility - will your design allow for delivery of potential future needs? Can you use a simpler framework that will allow you to deliver more quickly?

Reviewer 2: Is it worth mentioning keeping a decision log and other documentation? Or is that out of scope?

I suppose we can't include everything or it will be too long.

Reviewer: Could be worth mentioning, I think we _tried_ to get the outcomes first, then how you might achieve it, so something like the following maybe?
```suggestion
- Keep it simple, but think about extensibility - will your design allow for delivery of potential future needs? Can you use a simpler framework that will allow you to deliver more quickly?
- Keep track of decisions made, so progress is documented. This enabled others to follow the tracks and not spend time making the same choices.
```

Author: There is another principle coming along that talks about design and evidence, I think this might make more sense in there
